### DISCUSSION TITLE
Referrals

### DISCUSSION THREAD

**Discussion by Sohalt** [pretix/pretix#1966] [Ideas]
It would be cool to have the option of referring other users using a referral code or similar system. The referral code could act like a voucher, allowing access to certain products or giving a discount. The system should also save the referrer-referred relation, so it is possible to analyze who has referred how many people and reward people for referring others.

**Comment by sylvinus** [comment]
Hi!

We've done this for https://dotjs.io :)

Our method was to create a "template" voucher in the admin, and then duplicate if for each order to create a custom voucher to be given to their friends. We then analyze referrals manually, and we haven't implemented further relations.

hope this helps!

from `event.py`:

```

        # Set random referal codes for all attendees
        questions = {
            str(q.question): q
            for q in Question.objects.filter(event=event)
        }

        ref_template = Voucher.objects.filter(code="REFERAL_TEMPLATE", event=event)
        if len(ref_template) and questions.get("referal_code"):
            ref_template = ref_template[0]

            for pos in positions:

                initials = "".join([x[0].upper() for x in order["invoice_address"]["name"].split(" ") if len(x) > 0])
                new_code = "REF%s%s" % (initials, pos.id)

                # Save the code in the DB
                ref_template.pk = None
                ref_template.code = new_code
                ref_template.save()

                # Here we can safely call .add() because it must be the first answer.
                pos.answers.add(QuestionAnswer.objects.create(
                    orderposition=pos, question=questions["referal_code"], answer=new_code
                ))
```

**Comment by raphaelm** [comment]
Yeah, that sounds like a good option. You could implement that in a plugin that listens to order_placed. Instead of a question, you could also use OrderPosition.meta_info, if you want.
